pennylane/commutation_dag.py

```diff
@@ -0,0 +1,509 @@
+        "PauliX": 1,
+        "PauliY": 2,
+        "PauliZ": 3,
+        "SWAP": 4,
+        "ctrl": 5,
+    }
+)
+
+commutation_map = OrderedDict(
+    {
+        "Hadamard": [
+            1,
+            0,
+            0,
+            0,
+            0,
+            0,
+        ],
+        "PauliX": [
+            0,
+            1,
+            0,
+            0,
+            0,
+            0,
+        ],
+        "PauliY": [
+            0,
+            0,
+            1,
+            0,
+            0,
+            0,
+        ],
+        "PauliZ": [
+            0,
+            0,
+            0,
+            1,
+            0,
+            1,
+        ],
+        "SWAP": [
+            0,
+            0,
+            0,
+            0,
+            1,
+            0,
+        ],
+        "ctrl": [
+            0,
+            0,
+            0,
+            1,
+            0,
+            1,
+        ],
+    }
+)
+
+
+def intersection(wires1, wires2):
+    r"""Check if two operations are commuting
+
+    Args:
+        wires1 (pennylane.wires.Wires): First set of wires.
+        wires2 (pennylane.wires.Wires: Second set of wires.
+
+    Returns:
+         bool: True if the two sets of wires are not disjoint and False if disjoint.
+    """
+    if len(qml.wires.Wires.shared_wires([wires1, wires2])) == 0:
+        return False
+    return True
+
+
+def is_commuting(operation1, operation2):
+    r"""Check if two operations are commuting
+
+    Args:
+        operation1 (pennylane.Operation): A first quantum operation.
+        operation2 (pennylane.Operation): A second quantum operation.
+
+    Returns:
+         bool: True if the operations commute, False otherwise.
+
+    **Example**
+
+    >>> qml.is_commuting(qml.PauliX(wires=0), qml.PauliZ(wires=0))
+    True
+    """
+    # pylint: disable=too-many-branches
+    # pylint: disable=too-many-return-statements
+
+    # Case 1 operations are disjoints
+    if not intersection(operation1.wires, operation2.wires):
+        return True
+
+    # Case 2 both operations are controlled
+    if operation1.is_controlled and operation2.is_controlled:
+        control_control = intersection(operation1.control_wires, operation2.control_wires)
+        target_target = intersection(operation1.target_wires, operation2.target_wires)
+        control_target = intersection(operation1.control_wires, operation2.target_wires)
+        target_control = intersection(operation1.target_wires, operation2.control_wires)
+
+        # Case 2.1: disjoint targets
+        if control_control and not target_target and not control_target and not target_control:
+            return True
+
+        # Case 2.2: disjoint controls
+        if not control_control and target_target and not control_target and not target_control:
+            return bool(
+                commutation_map[operation1.is_controlled][position[operation2.is_controlled]]
+            )
+
+        # Case 2.3: targets overlap and controls overalp
```

Comment:
```suggestion
        # Case 2.3: targets overlap and controls overlap
```
the comments are really helpful for following the logic!